### PR TITLE
Abort test result available notification scheduling when viewed (EXPOSUREAPP-4516)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Abort test-result-available notification scheduling when result has been viewed
Renaming to make it easier to distinguish between share-result and result-available notification -> most changes are just renaming
Converted to draft so I can test it over the weekend

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/testresult/invalid/SubmissionTestResultInvalidViewModel.kt [hunk 1]
```diff
@@ -39,6 +41,11 @@ class SubmissionTestResultInvalidViewModel @AssistedInject constructor(
         }
     }
 
+    fun onTestOpened() {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/testresult/invalid/SubmissionTestResultInvalidViewModel.kt:44]: Persists that test result has been viewed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/testresult/negative/SubmissionTestResultNegativeViewModel.kt [hunk 2]
```diff
@@ -38,6 +40,11 @@ class SubmissionTestResultNegativeViewModel @AssistedInject constructor(
         }
     }
 
+    fun onTestOpened() {
+        submissionRepository.setViewedTestResult()
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/testresult/negative/SubmissionTestResultNegativeViewModel.kt:44]: Persists that test result has been viewed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/notification/NotificationConstants.kt [hunk 3]
```diff
@@ -18,7 +18,6 @@ object NotificationConstants {
 
     const val DEADMAN_NOTIFICATION_ID: NotificationId = 3
     const val NEW_MESSAGE_RISK_LEVEL_SCORE_NOTIFICATION_ID: NotificationId = 110
-    const val NEW_MESSAGE_TEST_RESULT_NOTIFICATION_ID: NotificationId = 120
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/notification/NotificationConstants.kt:21]: Why was this removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt [hunk 4]
```diff
@@ -376,18 +362,6 @@ object LocalData {
         false
     )
 
-    /**
-     * Toggles the boolean if the user has mobile data enabled
-     *
-     */
-    fun toggleMobileDataEnabled() = getSharedPreferenceInstance().edit(true) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt:383]: Why was this removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt [hunk 5]
```diff
@@ -354,18 +352,6 @@ object LocalData {
         false
     )
 
-    /**
-     * Toggles the decision if background jobs are enabled
-     *
-     */
-    fun toggleBackgroundJobEnabled() = getSharedPreferenceInstance().edit(true) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt:361]: Why was this removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt [hunk 6]
```diff
@@ -50,20 +52,16 @@ class DiagnosisTestResultRetrievalPeriodicWorker @AssistedInject constructor(
...
-            } else {
+
+            if (abortConditionsMet()) {
+                Timber.tag(TAG).d(" $id Stopping worker.")
                 stopWorker()
-                Timber.tag(TAG).d(" $id worker stopped")
             }
+
+            Timber.tag(TAG).d(" $id Running worker.")
+            val registrationToken = LocalData.registrationToken() ?: throw NoRegistrationTokenSetException()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt:62]: Doesn't this now get executed, even if `abortConditionsMet()` returns true?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt [hunk 7]
```diff
@@ -73,35 +71,57 @@ class DiagnosisTestResultRetrievalPeriodicWorker @AssistedInject constructor(
...
      * @see LocalData.initialPollingForTestResultTimeStamp
      * @see TestResult
      */
-    private suspend fun initiateNotification(testResult: TestResult) {
-        if (LocalData.isTestResultNotificationSent() || LocalData.submissionWasSuccessful()) {
-            Timber.tag(TAG).d("$id: Notification already sent or there was a successful submission")
-            return
-        }
-        Timber.tag(TAG).d("$id: Test Result retried is $testResult")
+    private suspend fun initiateTestResultAvailableNotification(testResult: TestResult) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt:108]: Maybe we could have this return a  `boolean` and then pull the `stopWorker()` up, so all `stopWorker()` calls are part of the top level function.

**Review by d4rken — COMMENTED** [review]
Fix looks good, but I think there might be a bug in the refactored worker.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt (hunk 5 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt:361]: Unused

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt (hunk 4 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt:383]: Unused

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt [hunk 8]
```diff
@@ -73,35 +71,57 @@ class DiagnosisTestResultRetrievalPeriodicWorker @AssistedInject constructor(
         return result
     }
 
+    private fun abortConditionsMet(): Boolean {
+        if (LocalData.isTestResultAvailableNotificationSent()) {
+            Timber.tag(TAG).d("$id: Notification already sent.")
+            return true
+        }
+        if (submissionSettings.hasViewedTestResult.value) {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt:79]: Check whether result has been viewed and abort

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/notification/NotificationConstants.kt (hunk 3 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/notification/NotificationConstants.kt:21]: Unused

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt (hunk 7 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorker.kt]: I moved some logic so that the stopWorker is on top level

**Review by d4rken — COMMENTED** [review]
Code looks good now, still need to test on device.

You edited the `Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorkerTest.kt`, did you also run it, because I think it's broken (instrumentation test)?

We can merge this after manual testing and get regression testing from testers & QA in the next RC, and then we should add unit tests for `DiagnosisTestResultRetrievalPeriodicWorker`,  in a follow up PR, what do you think @chiljamgossow ?

**Comment by ralfgehrer** [comment]
Code looks good. Device testing now.

**Comment by ralfgehrer** [comment]
Created a follow-up ticket for the missing unit tests: **EXPOSUREAPP-4663**

**Review by kolyaopahle — APPROVED** [review]
Worked on my device, code looks good 👍

**Review by ralfgehrer — APPROVED** [review]
